Podcast Player: Increase podcast shown

amustaque97 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->

Fixes #21291

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Created a new API to retrieve number of tracks.
* Added a method to call the above mentioned api
* Updating the max value of rangeControlUI

#### Jetpack product discussion NO
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request change what data or activity we track or use? NO
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->
* Checkout to this PR 
* Build and run project
* Create/edit a post and add `podcast player` block
* Use the following link to embed https://distributed.blog/category/podcast/feed/
* Add the following code in current theme `functions.php` file
```php
function my_jetpack_podcast_helper_tracks_quantity( $tracks_quantity ) {
	return 42;
}
add_filter( 'jetpack_podcast_helper_tracks_quantity', 'my_jetpack_podcast_helper_tracks_quantity' );

```
* Reload the page and check `number of items` button in podcast settings UI.
* Change value instead `return 15;` of filter to some different value and it will update max limit in the rangeControlUI as well.




**Screenshots**
- Number of Items
<img width="361" alt="image" src="https://user-images.githubusercontent.com/21127788/140568143-71a1c132-4496-4b82-857d-5e96f52470a5.png">

cc @samiff , I have created a separate branch because there are a lot of  changes in the previous branch.

amustaque97 committed:
Increase number of tracks when one sets up the PHP filter

This commit is a part of issue #21291. This commit is pushed in a
separate branch to avoid confusion from the previous suggested solution.
In this PR changes are:
* Introduced a new endpoint `track-quantity` to return number of tracks.
* Added a new method in the front end codebase to call the above mentioned API.
* After calling update `defaultMaxItems` of the rangeControlUI

amustaque97 committed:
Add Changelog

amustaque97 committed:
Merge remote-tracking branch 'origin' into add/increase-podcast-shown

samiff commented on projects/plugins/jetpack/_inc/lib/class-jetpack-podcast-helper.php:
```diff
@@ -25,6 +25,30 @@ public function __construct( $feed ) {
+		 * @param int    $number Number of tracks fetched. Default is 10.
+		 * @param object $rss    The SimplePie object built from core's `fetch_feed` call.
+		 */
+		return apply_filters( 'jetpack_podcast_helper_list_quantity', 10, $rss );
```
@jeherve The previous filter usage supplied the `$rss` SimplePie object, which makes it awkward when this filter is extracted out to a separate method.

What would you think about omitting the `$rss` arg from the `jetpack_podcast_helper_list_quantity` filter to simplify things? And _if_ it is removed, is there a certain annotation/practice for modifying filters like that? I can't imagine many would be using the `$rss` arg but 🤷

jeherve replied:
It's indeed not ideal. There are different ways we could go about this, but I think the safest would be to start deprecating `jetpack_podcast_helper_list_quantity`, and introduce a new filter without `$rss`. 

Something like this:

```diff
diff --git a/projects/plugins/jetpack/_inc/lib/class-jetpack-podcast-helper.php b/projects/plugins/jetpack/_inc/lib/class-jetpack-podcast-helper.php
index b23620d7ba..c79d047ce0 100644
--- a/projects/plugins/jetpack/_inc/lib/class-jetpack-podcast-helper.php
+++ b/projects/plugins/jetpack/_inc/lib/class-jetpack-podcast-helper.php
@@ -28,25 +28,18 @@ class Jetpack_Podcast_Helper {
 	/**
 	 * Retrieves track quantity.
 	 *
-	 * @returns Integer number of tracks
+	 * @returns int number of tracks
 	 */
 	public function get_track_quantity() {
-		$rss = $this->load_feed();
-
-		if ( is_wp_error( $rss ) ) {
-			return $rss;
-		}
-
 		/**
 		 * Allow requesting a specific number of tracks from SimplePie's `get_items` call.
 		 * The default number of tracks is ten.
 		 *
-		 * @since 9.5.0
+		 * @since 10.4.0
 		 *
-		 * @param int    $number Number of tracks fetched. Default is 10.
-		 * @param object $rss    The SimplePie object built from core's `fetch_feed` call.
+		 * @param int $number Number of tracks fetched. Default is 10.
 		 */
-		return apply_filters( 'jetpack_podcast_helper_list_quantity', 10, $rss );
+		return (int) apply_filters( 'jetpack_podcast_helper_tracks_quantity', 10 );
 	}
 
 	/**
@@ -200,6 +193,19 @@ class Jetpack_Podcast_Helper {
 
 		$tracks_quantity = $this->get_track_quantity();
 
+		/**
+		 * Allow requesting a specific number of tracks from SimplePie's `get_items` call.
+		 * The default number of tracks is ten.
+		 * Deprecated. Use jetpack_podcast_tracks_quantity filter instead, which takes one less parameter.
+		 *
+		 * @since 9.5.0
+		 * @deprecated 10.4.0
+		 *
+		 * @param int    $tracks_quantity Number of tracks fetched. Default is 10.
+		 * @param object $rss             The SimplePie object built from core's `fetch_feed` call.
+		 */
+		$tracks_quantity = apply_filters_deprecated( 'jetpack_podcast_helper_list_quantity', array( $tracks_quantity, $rss ), '10.4.0', 'jetpack_podcast_helper_tracks_quantity' );
+
 		// Process the requested number of items from our feed.
 		$track_list = array_map( array( __CLASS__, 'setup_tracks_callback' ), $rss->get_items( 0, $tracks_quantity ) );
 
```

amustaque97 replied:
Proposed changes make more sense to me and it will be easier to implement as well. We can start deprecating. 

One doubt is: Where is the filter definition `jetpack_podcast_helper_list_quantity`? I didn't find it in the codebase.

amustaque97 replied:
I will try proposed changes in my branch and will see if any new issue comes or not. In my opinion deprecating above fitler with `$rss` should not impact at least this issue.

samiff replied:
> Where is the filter definition jetpack_podcast_helper_list_quantity?

From what I see the filter only exists in that `class-jetpack-podcast-helper.php` file. If `add_filter()` is never called, then `apply_filters( $hook, $value )` will return `$value` by default. So it allows for hooking into the filter without needing an explicit `add_filter()` declaration. 

Let us know if any questions come up or when this is ready for another look.

amustaque97 replied:
@samiff, I have made the requested changes. thanks for clarifying `add_filter()`. It is working as expected.
Request you to please take an another round of look at the PR.

amustaque97 committed:
Implemented changes proposed by Jeremy, PHP Side

In this commit, we're deprecating `jetpack_podcast_helper_list_quantity` filter and introducing
a new filter `jetpack_podcast_helper_tracks_quantity`. New filter will work without $rss.

amustaque97 committed:
Wrapped tracks_quantity inside rest_ensure_response function.

amustaque97 committed:
Merge branch 'master' into add/increase-podcast-shown

samiff commented on projects/plugins/jetpack/extensions/blocks/podcast-player/edit.js:
```diff
@@ -194,6 +193,12 @@ const PodcastPlayerEdit = ( {
 			guids: selectedGuid ? [ selectedGuid ] : [],
 		} );
 
+		fetchTrackQuantity( {
```
Hmm this will run each time say `selectedGuid` changes, but really we should only need to fetch the `jetpack_podcast_helper_tracks_quantity` value once from that endpoint.

At the least I think it'd make sense moving the fetch to its own `useEffect`, but even then it'd be called say each time we insert a podcast player block on the same post.

I'm wondering if we could store that retrieved filter value as a piece of state somewhere? @jeherve do you have any suggestions?

amustaque97 replied:
I agree with you all. Overall I was also expecting same review comments and I wanted to push it this way only but now only thing i'm not sure of testing instruction. I will explain it below what i'm trying to convey here!

Since `get_tracks_quantity` is a static function, `add_filter` code will only work inside this function only. Because api endpoint is calling this function directly.
https://github.com/Automattic/jetpack/blob/dd22ab72fca0880ae68977319c26da71d1bc3ffb/projects/plugins/jetpack/_inc/lib/class-jetpack-podcast-helper.php#L33-L43

I was checking the docs about `add_filter` but couldn't find what is the best practice to write `add_filter` I mean whether we should write it in the class `__construct` or any class method and so on.

BTW Now I have updated the test instructions,  I would love to learn more about it.

samiff reviewed (changes requested):
Looking good and is working to do what we want 👍   Left some suggestions and a comment in hopes of polishing things up a bit.

amustaque97 committed:
Update comment to correct filter name

Co-authored-by: Samiff <samiff@users.noreply.github.com>

amustaque97 committed:
Define podcast get_tracks_quantity function static

Co-authored-by: Samiff <samiff@users.noreply.github.com>

amustaque97 committed:
Update track-quantity endpoint callback method to call static function

Co-authored-by: Samiff <samiff@users.noreply.github.com>

amustaque97 committed:
Update changelog

Co-authored-by: Samiff <samiff@users.noreply.github.com>

amustaque97 committed:
Address review comments.

In this commit, I have removed url parameter from fetchTrackQuantity api method and
I have written a separate hook to call the same function.

amustaque97 commented on projects/plugins/jetpack/extensions/blocks/podcast-player/api.js:
```diff
@@ -9,6 +9,15 @@ import { PODCAST_FEED, EMBED_BLOCK } from './constants';
 import apiFetch from '@wordpress/api-fetch';
 import { addQueryArgs } from '@wordpress/url';
 
+export const fetchTrackQuantity = async ( { url } ) => {
```
I have refactored this. Now `fetchTrackQuantity` doesn't take any parameter.

samiff commented on projects/plugins/jetpack/extensions/blocks/podcast-player/edit.js:
```diff
@@ -197,6 +196,13 @@ const PodcastPlayerEdit = ( {
 		return () => cancellableFetch?.current?.cancel?.();
 	}, [ fetchFeed, checkUrl, selectedGuid ] );
 
+	// Retrieve tracks quantity on initial load
```
```suggestion
	// Retrieve tracks quantity to fetch; the jetpack_podcast_helper_tracks_quantity filter value.
```

samiff commented on projects/plugins/jetpack/extensions/blocks/podcast-player/api.js:
```diff
@@ -9,6 +9,13 @@ import { PODCAST_FEED, EMBED_BLOCK } from './constants';
 
+export const fetchTrackQuantity = async () => {
+	const trackQuantity = await apiFetch( {
+		path: addQueryArgs( '/wpcom/v2/podcast-player/track-quantity' ),
```
```suggestion
		path: '/wpcom/v2/podcast-player/track-quantity'
```

samiff reviewed (changes requested):
Thanks for adding those changes, this is looking good! I added two minor suggestions, and please merge the master branch once more into your branch so the checks are able to pass.

> I was checking the docs about add_filter but couldn't find what is the best practice to write add_filter I mean whether we should write it in the class __construct or any class method and so on.

Hmm not sure I fully understood the question being asked here so feel free to clarify if I missed something. As an end user, I would be able to do something like

```php
function my_jetpack_podcast_helper_tracks_quantity( $tracks_quantity ) {
	return 42;
}
add_filter( 'jetpack_podcast_helper_tracks_quantity', 'my_jetpack_podcast_helper_tracks_quantity' );
```

My personal preference would be to add that to a `mu-plugin` file, or perhaps a child theme's functions.php file. Adding that example snippet seems to work well:

![Markup on 2021-11-17 at 12:47:24](https://user-images.githubusercontent.com/15803018/142271923-720b79f2-991a-45b5-80ca-fc0e32cc025d.png)

amustaque97 committed:
Remove addQueryArgs not required

Co-authored-by: Samiff <samiff@users.noreply.github.com>

amustaque97 committed:
Update comment to be more descriptive in podcast-player editjs

Co-authored-by: Samiff <samiff@users.noreply.github.com>

amustaque97 committed:
Merge remote-tracking branch 'origin/master' into add/increase-podcast-shown

amustaque97 commented:
PR is up to date 🎉  

Oh God! I was writing `add_filter` statement in the wp core `functions.php`! So stupid 😞 

Thanks once again Samiff & jeherve. Really appreciate your efforts ❤️ 

Hopefully, I will keep this in mind that there is `functions.php` file in theme as well. 😥

samiff commented:
Thanks for the contribution @amustaque97 these changes are now merged and will be shipped as part of Jetpack version 10.4 on 2021-12-07 🎉

samiff commented:
Internal: r235233-wpcom , D70255-code

inaikem commented:
**Edit:** I posted this request in the original issue here:
https://github.com/Automattic/jetpack/issues/21291
